stdlib/public/runtime/ProtocolConformance.cpp

```diff
@@ -1271,5 +1276,87 @@ const Metadata *swift::findConformingSuperclass(
   return conformingType;
 }
 
+SWIFT_CC(swift)
+void swift::swift_enumerateConformingTypesFromImage(
+  const void *imageAddress,
+  Metadata const *protocolMetadata,
+  TypeEnumerationFunction body,
+  void *bodyContext,
+  SWIFT_CONTEXT void *context,
+  SWIFT_ERROR_RESULT SwiftError **error) {
+
+  // Find the protocol(s) requested by the caller.
+  auto existentialType = dyn_cast<ExistentialTypeMetadata>(protocolMetadata);
+  if (!existentialType) {
+    auto typeName = swift_getTypeName(protocolMetadata, true);
+    swift::fatalError(
+        0,
+        "_enumerateTypes(fromImageAt:conformingTo:) can only test for "
+        "conformance to a protocol, but %.*s is not a protocol.",
+				(int)typeName.length, typeName.data);
+  }
+  const auto& protocols = existentialType->getProtocols();
+
+#if SWIFT_OBJC_INTEROP
+  // Check for any Objective-C protocols in the input. We do not currently
+  // support searching for Objective-C protocol conformances because they do not
+  // produce conformance records in the compiled binary. In the future, if we
+  // detect any such protocols, we can use objc_copyClassList() to add to the
+  // enumerated set (taking care not to enumerate a class twice if it conforms
+  // to an Objective-C protocol *and* a Swift protocol.)
+  for (const auto& protocol : protocols) {
+    if (protocol.isObjC()) {
+      swift::fatalError(
+          0,
+          "_enumerateTypes(fromImageAt:conformingTo:) does not currently "
+          "support Objective-C protocols such as %s.",
+          protocol.getName());
+    }
+  }
+#endif
+
+  // FIXME: capture image address at registration time so we don't need to look them up here?
+  auto snapshot = Conformances.get().SectionsToScan.snapshot();
+  for (const auto& section : snapshot) {
+
+    SymbolInfo info;
+    if (imageAddress) {
+      const auto sectionAddress = section.address();
+      if (!lookupSymbol(sectionAddress, &info)) {
+        // Couldn't find the image this section came from.
```

Comment:
Is this worth logging some sort of error around? Genuinely not sure how likely this is and mostly want to have extra logging if we have a weird scenario, more than anything.

Reply:
I hope this bit of code is temporary until I sort out a more efficient way to get the conformance section from an image address.

On Apple, Linux, and Windows, `lookupSymbol()` shouldn't fail in practice when passed a valid mapped pointer, and `sectionAddress` can only become invalid if somebody's pulling some `mmap()` shenanigans. On the other hand, at least some target platforms do not implement `lookupSymbol()` (e.g. WASM) and it always returns `false` for those platforms. So we can't just `assert()`.

Reply:
I refactored this code entirely.